cmd/report.go

```diff
@@ -0,0 +1,100 @@
+package cmd
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"net/http"
+	"runtime"
+	"strings"
+
+	"go.k6.io/k6/execution"
+	"go.k6.io/k6/lib/consts"
+)
+
+type report struct {
+	Version    string         `json:"k6_version"`
+	Executors  map[string]int `json:"executors"`
+	VUsMax     int64          `json:"vus_max"`
+	Iterations uint64         `json:"iterations"`
+	Duration   string         `json:"duration"`
+	GoOS       string         `json:"goos"`
+	GoArch     string         `json:"goarch"`
+	Modules    []string       `json:"modules"`
+	Outputs    []string       `json:"outputs"`
+}
+
+func createReport(execScheduler *execution.Scheduler, importedModules []string, outputs []string) report {
+	executors := make(map[string]int)
+	for _, ec := range execScheduler.GetExecutorConfigs() {
+		executors[ec.GetType()]++
+	}
+
+	// collect the report only with k6 public modules
+	publicModules := make([]string, 0, len(importedModules))
+	for _, module := range importedModules {
+		// Exclude JS modules extensions to prevent to leak
+		// any user's custom extensions
+		if strings.HasPrefix(module, "k6/x") {
+			continue
+		}
+		// Exclude any import not starting with the k6 prefix
+		// that identifies a k6 built-in stable or experimental module.
+		// For example, it doesn't include any modules imported from the file system.
+		if !strings.HasPrefix(module, "k6") {
+			continue
+		}
+		publicModules = append(publicModules, module)
+	}
+
+	builtinOutputs := map[string]bool{
+		"cloud": true, "csv": true, "experimental-prometheus-rw": true,
+		"influxdb": true, "json": true, "statsd": true,
+	}
+
+	// collect only the used outputs that are builtin
+	publicOutputs := make([]string, 0, len(outputs))
+	for _, o := range outputs {
+		if !builtinOutputs[o] {
+			continue
+		}
+		publicOutputs = append(publicOutputs, o)
+	}
+
+	execState := execScheduler.GetState()
+	return report{
+		Version:    consts.Version,
+		Executors:  executors,
+		VUsMax:     execState.GetInitializedVUsCount(),
+		Iterations: execState.GetFullIterationCount(),
+		Duration:   execState.GetCurrentTestRunDuration().String(),
+		GoOS:       runtime.GOOS,
+		GoArch:     runtime.GOARCH,
+		Modules:    publicModules,
+		Outputs:    publicOutputs,
+	}
+}
+
+func reportUsage(ctx context.Context, execScheduler *execution.Scheduler, test *loadedAndConfiguredTest) error {
+	outputs := make([]string, 0, len(test.derivedConfig.Out))
+	for _, o := range test.derivedConfig.Out {
+		outputName, _ := parseOutputArgument(o)
+		outputs = append(outputs, outputName)
+	}
+
+	r := createReport(execScheduler, test.moduleResolver.Imported(), outputs)
+	body, err := json.Marshal(r)
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, "https://reports.k6.io", bytes.NewBuffer(body))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/json")
+	res, err := http.DefaultClient.Do(req)
```

Comment:
I know the use of the `DefaultClient` is generally discouraged, because it lacks the definition of timeouts and other proper settings for production environments. Is it safe to use it here?

Reply:
>it lacks the definition of timeouts
>Is it safe to use it here?

Specifically, regarding the timeout, we should be safe as we are passing the timeout-limited context.

In general, it would be a good potential improvement. However, not so far in the future, we have to migrate to the new Grafana's service API. Then it may be convenient to rewrite and refactor this client all at once.

Reply:
Fair enough, let's move on! Also, looks like this code already existed before, so 🚀